### PR TITLE
[FLINK-24651][iteration] Add bounded all-round iteration

### PR DESCRIPTION
**Pull request by gaoyunhaii**
This PR adds the support of the bounded all-round iteration.

### CODE REVIEW

File: flink-ml-iteration/src/main/java/org/apache/flink/iteration/IterationFactory.java [hunk 1]
```diff
@@ -109,11 +111,78 @@ public static DataStreamList createIteration(
...
+            DraftExecutionEnvironment draftEnv,
+            DataStreamList initVariableStreams,
+            DataStreamList headStreams,
+            int totalInitVariableParallelism) {
+        // deal with the criteria streams
+        DataStream<?> terminationCriteria = draftEnv.getActualStream(draftCriteriaStream.getId());
+        // It should always has the IterationRecordTypeInfo
+        checkState(
+                terminationCriteria.getType().getClass().equals(IterationRecordTypeInfo.class),
+                "The termination criteria should always returns IterationRecord.");
```
Comment (guoweiM) [inline flink-ml-iteration/src/main/java/org/apache/flink/iteration/IterationFactory.java:141]: returns --> return

File: flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            }
+
+            verifyResult(roundsStat, 5, 1, 4 * (0 + 999) * 1000 / 2);
+            assertEquals(OutputRecord.Event.TERMINATED, result.take().getEvent());
+        }
+    }
+
+    @Test
+    public void testTerminationCriteria() throws Exception {
+        try (MiniCluster miniCluster = new MiniCluster(createMiniClusterConfiguration(2, 2))) {
```
Comment (guoweiM) [inline flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java:148]: see above

File: flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+    private static BlockingQueue<OutputRecord<Integer>> result = new LinkedBlockingQueue<>();
+
+    @Before
+    public void setup() {
+        result.clear();
+    }
+
+    @Test(timeout = 60000)
+    public void testSyncVariableOnlyBoundedIteration() throws Exception {
+        try (MiniCluster miniCluster = new MiniCluster(createMiniClusterConfiguration(2, 2))) {
```
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java:70]: Similarly I extract the process of minicluster

File: flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+import static org.apache.flink.iteration.itcases.UnboundedStreamIterationITCase.verifyResult;
+import static org.junit.Assert.assertEquals;
+
+/**
+ * Tests the cases of {@link Iterations#iterateBoundedStreamsUntilTermination(DataStreamList,
+ * DataStreamList, IterationBody)}.
+ */
+public class BoundedAllRoundStreamIterationITCase {
+
+    private static BlockingQueue<OutputRecord<Integer>> result = new LinkedBlockingQueue<>();
```
[the remark this answers is not in the document]
Comment (gaoyunhaii) [reply flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java:61]: I also change this test to SharedObject

File: flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (guoweiM) [reply flink-ml-iteration/src/test/java/org/apache/flink/iteration/itcases/BoundedAllRoundStreamIterationITCase.java]: OK. But I think you could create a jira to track this improvement.

**Review by guoweiM — APPROVED** [review]

**Comment by gaoyunhaii** [comment]
Very thanks @guoweiM  for the review! will merge~

**Pull request closed without merging**